### ISSUE TITLE
Error when downloading new torrents

### ISSUE THREAD

**Issue by escape0707** [arvidn/libtorrent#6612]
Qbittorrent 4.4rc1 built from AUR
Libtorrent 2.0.5 built from AUR
Archlinux 5.15.11
Ntfs3

When I'm downloading a torrent into a ntfs3 mounted external hdd drive, qbit always first create those files waiting to be downloaded, then it fails and spit a log with:

`2021/12/03 15:47 - File error alert. Torrent: "xxx". File: "/run/media/jay/Seagate Basic/my-files/Downloads/Bittorrent/xxx/yyy.mkv". Reason: xxx file_open (/run/media/jay/Seagate Basic/my-files/Downloads/Bittorrent/yyy.mkv) error: resource temporarily unavailable`

If I try resume several times, it will retry and fails then finally finish creating those files and continue to download.

When using the same setup to download to $HOME which is on the internal ssd and formatted as ext4, there is no problem.

I run into this problem even when downloading a torrent with only 9 files.

**Comment by arvidn** [comment]
I believe this is a bug (or missing feature) in ntfs-3g. libtorrent supports a disk I/O back-end that doesn't use memory mapped files (which is the default in 32 bit builds). I don't know if qbt lets you pick this though, but it could.

It would have to use the [posix_disk_io_constructor](http://libtorrent.org/reference-Storage.html#posix_disk_io_constructor()). That might fix it. I'm suspecting the ntfs filesystem is having a hard time with memory mapped files.

**Comment by escape0707** [comment]
I'm using Linux kernel's new ntfs3 implementation, not the NTFS-3G one. I think NTFS-3G works, but I'm not sure. Should I try it out?

**Comment by arvidn** [comment]
I see, I just assumed it was ntfs-3g you were using.

What's puzzling is that the man page for `open()` doesn't list `EAGAIN` as a possible error.

Looking over the code, there seems to be an error where the true `operation` field may get clobbered by `file_open`. Would it be possible for you to test this patch? https://github.com/arvidn/libtorrent/pull/6617

If so, the error message may be a bit more helpful.

**Comment by arvidn** [comment]
it seems likely that it's the `mmap()` call that's failing with `EAGAIN` ([man page](https://man7.org/linux/man-pages/man2/mmap.2.html)).

> EAGAIN The file has been locked, or too much memory has been
> locked (see setrlimit(2)).

**Comment by escape0707** [comment]
New error log:

```
2021/12/29 12:58 - File error alert. Torrent: "xxx". File: "/run/media/jay/Seagate Basic/my-files/Downloads/Bittorrent/xxx/yyy.mkv". Reason: xxx file_mmap (/run/media/jay/Seagate Basic/my-files/Downloads/Bittorrent/yyy.mkv) error: resource temporarily unavailable
```

**Comment by escape0707** [comment]
Should I provide any else info?

**Comment by escape0707** [comment]
This is still happening when using qBt 4.4.

**Comment by arvidn** [comment]
something in the ntfs filesystem driver is failing with `EAGAIN`:

> The file has been locked, or too much memory has been
> locked (see setrlimit(2)).

So, perhaps it's causing memory to be locked exceeding your ulimit. To me, it smells like a quality-of-implementation issue in the ntfs filesystem driver.

**Comment by database64128** [comment]
Besides the `Resource temporarily unavailable` errors when creating files, I'm also seeing system lockups during initial file writes. Stats in `btm`/`htop` would stop updating and qBittorrent would stop responding for an extended period of time, until a large chunk of zeros is written to disk. It seems to be partially caused by the fact that NTFS can't "store several unwritten areas": https://github.com/torvalds/linux/blob/719fce7539cd3e186598e2aed36325fe892150cf/fs/ntfs3/file.c#L690-L700.

I did a little experiment to compare `ntfs3` to other file systems. Create an empty 10G file, seek to 8G and write 4K:

```
truncate -s 10G test
dd if=/dev/urandom of=test bs=4K count=1 seek=2M status=progress
```

On `btrfs`, both `truncate` and `dd` operations complete immediately and barely any data is written to disk.

```
1+0 records in
1+0 records out
4096 bytes (4.1 kB, 4.0 KiB) copied, 0.000123942 s, 33.0 MB/s
```

On `ntfs3`, `truncate` completes immediately. `dd` took 35s. `ntfs3` basically zeroed out the entire 8G region before the actual data.

```
4096 bytes (4.1 kB, 4.0 KiB) copied, 35 s, 0.1 kB/s
1+0 records in
1+0 records out
4096 bytes (4.1 kB, 4.0 KiB) copied, 35.4918 s, 0.1 kB/s
```

**Comment by escape0707** [comment]
> I'm also seeing system lockups during initial file writes. Stats in btm/htop would stop updating and qBittorrent would stop responding for an extended period of time

Qbittorrent itself is locked up, too.

**Comment by database64128** [comment]
`ntfs3` maintainer just sent a pull request with a bunch of fixes for 5.19. Keep this issue open for further updates.

**Comment by afeuerstein** [comment]
I can reproduce on 5.18.6-artix Artix Linux w/ open-rc.

```
An I/O error occurred for torrent 'torrent'.
Reason: file_mmap (/ntfs/torrent/) error: Resource temporarily unavailable```
